File: core/geology.py
"""
Geekatplay Studio — Structural Geology & Tectonic Solvers.
Simulates geological rock strata, differential lithology hardness, tectonic faults, and plateaus.
"""

import numpy as np
from typing import Tuple


def geological_strata(
    heightfield: np.ndarray,
    layers_count: int = 12,
    hardness_variance: float = 0.6,
    folding_angle_deg: float = 15.0,
    seed: int = 42
) -> np.ndarray:
    """
    Applies multi-layered geological sedimentary strata with differential erosion resistance.
    Creates realistic cliff ledges, stepped plateau canyon walls, and rock banding.
    """
    h = heightfield.astype(np.float32).copy()
    res_y, res_x = h.shape
    rng = np.random.default_rng(seed)

    # Hardness profiles for distinct strata bands
    layer_hardness = rng.uniform(1.0 - hardness_variance, 1.0 + hardness_variance, size=layers_count)

    # Compute strata phase including dip angle
    rad = np.radians(folding_angle_deg)
    lin_x = np.linspace(0.0, 1.0, res_x, dtype=np.float32)
    lin_y = np.linspace(0.0, 1.0, res_y, dtype=np.float32)
    grid_x, grid_y = np.meshgrid(lin_x, lin_y)
    
    dip_offset = (grid_x * np.cos(rad) + grid_y * np.sin(rad)) * 0.15
    strata_coord = np.clip((h + dip_offset) * layers_count, 0.0, layers_count - 1.001)

    layer_idx = np.floor(strata_coord).astype(int)
    frac = strata_coord - layer_idx

    # S-curve terrace step shaping based on layer hardness
    hardness = layer_hardness[layer_idx]
    shaped_frac = np.where(hardness > 1.0, np.power(frac, 0.5), np.power(frac, 2.0))

    strata_h = (layer_idx + shaped_frac) / layers_count
    blended = h * 0.7 + strata_h * 0.3
    return np.clip(blended, 0.0, 1.0).astype(np.float32)


def tectonic_fault(
    heightfield: np.ndarray,
    fault_angle_deg: float = 45.0,
    displacement_height: float = 0.25,
    fault_position: float = 0.5,
    drag_width: float = 0.08,
) -> np.ndarray:
    """
    Simulates tectonic fault displacement (normal / reverse faulting).
    Creates sheer escarpments, fault rifts, and tectonic mountain uplifts.
    """
    h = heightfield.astype(np.float32).copy()
    res_y, res_x = h.shape

    rad = np.radians(fault_angle_deg)
    nx = np.cos(rad)
    ny = np.sin(rad)

    lin_x = np.linspace(0.0, 1.0, res_x, dtype=np.float32)
    lin_y = np.linspace(0.0, 1.0, res_y, dtype=np.float32)
    gx, gy = np.meshgrid(lin_x, lin_y)

    dist_from_fault = (gx * nx + gy * ny) - fault_position

    # Smooth sigmoid transition along the fault drag zone
    displacement = displacement_height * (1.0 / (1.0 + np.exp(-dist_from_fault / max(drag_width, 1e-4))))
    uplifted = h + displacement

    min_val = np.min(uplifted)
    max_val = np.max(uplifted)
    if max_val > min_val:
        uplifted = (uplifted - min_val) / (max_val - min_val)
    return uplifted.astype(np.float32)


def terrace_plateau(
    heightfield: np.ndarray,
    steps_count: int = 6,
    smoothness: float = 0.15
) -> np.ndarray:
    """Carves terraced plateau steps with customizable cliff sharpness."""
    h = heightfield.astype(np.float32)
    scaled = h * steps_count
    base = np.floor(scaled)
    frac = scaled - base

    # Smoothstep interpolation
    t = np.clip((frac - (0.5 - smoothness)) / (2.0 * smoothness + 1e-6), 0.0, 1.0)
    stepped = base + t * t * (3.0 - 2.0 * t)
    return (stepped / steps_count).astype(np.float32)

File: src/geology.cpp
#include "../include/geekatplay/geology.hpp"
#include <cmath>
#include <random>
#include <algorithm>

namespace Geekatplay {

std::vector<float> GeologySolver::GeologicalStrata(
    const std::vector<float>& inH, int res, int layers, float hardnessVariance, float foldingAngleDeg
) {
    std::vector<float> outH = inH;
    float rad = foldingAngleDeg * 3.1415926535f / 180.0f;
    float cosA = std::cos(rad);
    float sinA = std::sin(rad);

    for (int y = 0; y < res; ++y) {
        for (int x = 0; x < res; ++x) {
            int idx = y * res + x;
            float gx = static_cast<float>(x) / res;
            float gy = static_cast<float>(y) / res;

            float dipOffset = (gx * cosA + gy * sinA) * 0.15f;
            float strataCoord = std::clamp((inH[idx] + dipOffset) * layers, 0.0f, static_cast<float>(layers - 1));

            int layerIdx = static_cast<int>(std::floor(strataCoord));
            float frac = strataCoord - layerIdx;

            // S-curve terrace profiling modulated by hardnessVariance
            float expLow = std::max(0.2f, 1.0f - hardnessVariance * 0.6f);
            float expHigh = 1.0f + hardnessVariance * 0.8f;
            float shapedFrac = (layerIdx % 2 == 0) ? std::pow(frac, expLow) : std::pow(frac, expHigh);
            float strataH = (layerIdx + shapedFrac) / layers;

            outH[idx] = std::clamp(inH[idx] * 0.7f + strataH * 0.3f, 0.0f, 1.0f);
        }
    }
    return outH;
}

std::vector<float> GeologySolver::TectonicFault(
    const std::vector<float>& inH, int res, float faultAngleDeg, float displacement, float faultPosition
) {
    std::vector<float> outH = inH;
    float rad = faultAngleDeg * 3.1415926535f / 180.0f;
    float nx = std::cos(rad);
    float ny = std::sin(rad);

    for (int y = 0; y < res; ++y) {
        for (int x = 0; x < res; ++x) {
            int idx = y * res + x;
            float gx = static_cast<float>(x) / res;
            float gy = static_cast<float>(y) / res;

            float dist = (gx * nx + gy * ny) - faultPosition;
            float disp = displacement * (1.0f / (1.0f + std::exp(-dist / 0.08f)));
            outH[idx] += disp;
        }
    }

    float minVal = *std::min_element(outH.begin(), outH.end());
    float maxVal = *std::max_element(outH.begin(), outH.end());
    if (maxVal > minVal) {
        for (auto& v : outH) {
            v = (v - minVal) / (maxVal - minVal);
        }
    }
    return outH;
}

std::vector<float> GeologySolver::TerracePlateau(
    const std::vector<float>& inH, int res, int steps, float smoothness
) {
    std::vector<float> outH(res * res, 0.0f);
    for (int i = 0; i < res * res; ++i) {
        float scaled = inH[i] * steps;
        float base = std::floor(scaled);
        float frac = scaled - base;

        float t = std::clamp((frac - (0.5f - smoothness)) / (2.0f * smoothness + 1e-6f), 0.0f, 1.0f);
        float stepped = base + t * t * (3.0f - 2.0f * t);
        outH[i] = stepped / steps;
    }
    return outH;
}

} // namespace Geekatplay
